### PR TITLE
fixed deadlocks, semaphores deserialization regarding the comments, also css specificity

### PR DESCRIPTION
**Pull request by nickdaminov**
# Done:

  - [x] changed based on comments
  - [x] added some css specificity

# To do:

- [ ] need  review

### CODE REVIEW

**Comment by nickdaminov** [comment]
@mihaiberq pls review

File: Java/Application Integrity/synchronization-i/semaphores.md [hunk 1]
```diff
@@ -16,7 +16,7 @@ links:
 
 Semaphores are very similar to locks, the only difference is that they allow multiple threads/processes to enter the *critical section*[1]
```
Comment (mihaiberq) [inline Java/Application Integrity/synchronization-i/semaphores.md]: `.` here

File: Java/Application Integrity/synchronization-i/semaphores.md [hunk 2]
```diff
@@ -16,7 +16,7 @@ links:
...
 
-In the following example *semaphore* class keeps track of how many threads entered it and only allows new threads to enter if their
+The following code is an example of a basic *semaphore* class. It keeps track of how many threads have entered the semaphore and only allows new threads to enter if the limit is not reached.
```
Comment (mihaiberq) [inline Java/Application Integrity/synchronization-i/semaphores.md]: have acquired the semaphore

File: CSS/HTML and CSS Basics/css-basics/css-specificity.md [hunk 3]
```diff
@@ -28,96 +28,73 @@ links:
...
-```html
-<div>
-  <div class="my-class">
-    <h2 id="title"
-      style="font-size:20px;">
-      Title
-    </h2>
-  </div>
-</div>
+In this case *id "dream-garage"* selector is more **specific** than *class "favourite"* which neutralises effect of the latter.
```
Comment (mihaiberq) [inline CSS/HTML and CSS Basics/css-basics/css-specificity.md]: In this case, the id selector "dream-garage" is more specific than the "favourite" class, which effectively overrides the effect of the later.

File: CSS/HTML and CSS Basics/css-basics/css-specificity.md [hunk 4]
```diff
@@ -28,96 +28,73 @@ links:
...
+**Specificity** determines which selector affects desired piece of *HTML* code, this goes in addition to *cascading*. Let's take a look at four different selector attributes  and the *CSS* inline styling:
```
Comment (mihaiberq) [inline CSS/HTML and CSS Basics/css-basics/css-specificity.md]: of html code and this goes in addition to cascading.

File: CSS/HTML and CSS Basics/css-basics/css-specificity.md [hunk 5]
```diff
@@ -28,96 +28,73 @@ links:
...
+ - Inline styles
+ - IDs
+ - Classes, attributes and pseudo-classes
+ - classesElements and pseudo-elements
+
+The ones at the bottom have the lowest *specificity* and are overridden by the styling under higher attributes. The way to calculate the *specificity* of a selector is to calculate the amount of each of the above attributes, the one with most of them wins.
+
```
Comment (mihaiberq) [inline CSS/HTML and CSS Basics/css-basics/css-specificity.md]: To put everything in terms of cascading (about which you should remember that styles defined later in the CSS file are the ones used): if selector A has a higher specificity than selector B, with A and B both defined in the same CSS file, A would effectively be placed below B.

File: CSS/HTML and CSS Basics/css-basics/css-specificity.md [hunk 6]
```diff
@@ -28,96 +28,73 @@ links:
...
+Example:
 ```
-We would have to add a new column to the previous notation in order to represent inline CSS specificity:
-```text
-internal - id - class - type
-    1    - 0  -   0   -  0
+.favorite: 0 inline, 0 ID, 1 Class, 0 Elem
+ in short: 0-0-1-0
+
+ul#dream-garage li: 0 inline, 1 ID,
```
Comment (mihaiberq) [inline CSS/HTML and CSS Basics/css-basics/css-specificity.md]: same comment about ul

**Comment by nickdaminov** [comment]
@stefkn pls review

File: CSS/HTML and CSS Basics/css-basics/css-specificity.md (hunk 5 above)
Comment (lizTheDeveloper) [reply CSS/HTML and CSS Basics/css-basics/css-specificity.md]: I liked the tables that were there, they made the concept very clear. Can you put those back?

**Comment by Loopiezlol** [comment]
@nickdaminov any update on this?

**Pull request merged**